data-science-onramp/data-ingestion/setup-test.py

```diff
@@ -0,0 +1,210 @@
+import os
+import re
+
+import uuid
+
+from google.api_core.exceptions import GoogleAPICallError
+
+from google.cloud import dataproc_v1 as dataproc
+from google.cloud import storage
+from google.cloud.exceptions import NotFound
+
+import pytest
+
+waiting_cluster_callback = False
+
+# Set global variables
+project = os.environ['GCLOUD_PROJECT']
+region = "us-central1"
+zone = "us-central1-a"
+cluster_name = 'setup-test-{}'.format(str(uuid.uuid4()))
+bucket_name = 'setup-test-code-{}'.format(str(uuid.uuid4()))
+
+
+@pytest.fixture(autouse=True)
+def teardown():
+    yield
+
+    # Delete cluster
+    cluster_client = dataproc.ClusterControllerClient(client_options={
+        'api_endpoint': f'{region}-dataproc.googleapis.com:443'
+    })
+
+    try:
+        operation = cluster_client.delete_cluster(project, region,
+                                                  cluster_name)
+        operation.result()
+    except GoogleAPICallError:
+        pass
+
+    # Delete GCS bucket
+    storage_client = storage.Client()
+    try:
+        bucket = storage_client.get_bucket(bucket_name)
+        bucket.delete(force=True)
+    except NotFound:
+        pass
+
+
+def test_setup(capsys):
```

Comment:
All of your setup code should go into your `teardown` function above (perhaps renamed to something like "setup_teardown". Based on wherever you put the `yield`, all code above it will run BEFORE executing the test, and all code after it will run AFTER. See here: [dataproc_quickstart](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/master/dataproc/quickstart/quickstart_test.py)

In your code, I would create the cluster and GCS buckets in your setup/teardown function. I would leave the test function itself to only include elements of the test that involve submitting the actual job.

Reply:
ah yep - I see Brad called this out already 👍